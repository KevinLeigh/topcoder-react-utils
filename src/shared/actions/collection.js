/**
 * Proof-of-concept code. Will be documented once stable. In case of any
 * questions before that, feel free to contact `birdofpreyru`.
 */

import _ from 'lodash';
import * as redux from 'utils/redux';

/**
 * Adds items to the collection.
 * @param {Object} index Item ID > item mapping for the items to add.
 * @return {Object}
 */
function addItems(index) {
  return { index, timestamp: Date.now() };
}

/**
 * Increments counters of item references.
 * @param {String|String[]} ids
 * @return {String[]}
 */
function bookItems(ids) {
  return _.isArray(ids) ? ids : [ids];
}

/**
 * Removes from the state all unuzed items with timestamps older than maxage.
 * @param {Number} timestamp
 */
function clean(timestamp) {
  return timestamp;
}

/**
 * Decreases counters of item references.
 * @param {String|String[]} ids
 * @return {String[]}
 */
function freeItems(ids) {
  return _.isArray(ids) ? ids : [ids];
}

/**
 * Inits loading of an item.
 * @param {String} operationId
 * @param {String} itemId
 * @return {Object}
 */
function loadItemInit(operationId, itemId) {
  return { itemId, operationId };
}

/**
 * Finalizes loading of an item.
 * @param {String} operationId
 * @param {String} itemId
 * @param {Any} item
 * @return {Object}
 */
function loadItemDone(operationId, itemId, item) {
  return {
    item,
    itemId,
    operationId,
    timestamp: Date.now(),
  };
}

export default redux.createActions({
  COLLECTION: {
    ADD_ITEMS: addItems,
    BOOK_ITEMS: bookItems,
    CLEAN: clean,
    FREE_ITEMS: freeItems,
    LOAD_ITEM_INIT: loadItemInit,
    LOAD_ITEM_DONE: loadItemDone,
  },
});
