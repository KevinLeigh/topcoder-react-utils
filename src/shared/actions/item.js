/**
 * @module "actions.item"
 * @desc Actions for management of item data in Redux store.
 */

import * as redux from 'utils/redux';

/**
 * @static
 * @desc Creates an action that drops item's data if (i) the value of item's
 *  reference counter is not positive; (ii) item's timestamp is older than
 *  `olderThan` value.
 * @param {Number} olderThan=Number.MAX_VALUE Optional. Timestamp [ms].
 *  If provided, item's  data will be dropped only if they were received / set
 *  prior to this moment in time.
 * @return {Action}
 */
function dropData(olderThan = Number.MAX_VALUE) {
  return olderThan;
}

/**
 * @static
 * @desc Creates an action that signals the beginning of async loading of
 *  item's data.
 * @param {Number|String} loadingOperationId The unique identifier of the
 *  loading operation. It will be written into item's segment of Redux store
 *  with the following efffects:
 *
 *  - We consider that item's data are being loaded if, and only if,
 *    the item's `loadingOperationId` has a truly value;
 *
 *  - `loadingOperationId` is used to match the corresponding
 *    {@link module:actions.item.loadDataDone} operation.
 *
 * @return {Action}
 */
function loadDataInit(loadingOperationId) {
  return loadingOperationId;
}

/**
 * @static
 * @desc Creates an action that handles the result of async loading of item's
 *  data.
 * @param {Number|String} loadingOperationId The unique identifier of the
 *  loading operation. If it does not match the value currently written for
 *  the item in the Redux store (see {@link module:actions.item.loadDataInit}),
 *  the action will be silently ignored.
 * @param {Any} data Any value to store as the item's data.
 * @return {Action}
 */
function loadDataDone(loadingOperationId, data) {
  return { data, loadingOperationId, timestamp: Date.now() };
}

/**
 * @static
 * @desc Creates an action that stores given data into the item.
 * @param {Any} data Any value to store as the item's data.
 * @return {Action}
 */
function setData(data) {
  return { data, timestamp: Date.now() };
}

/**
 * @static
 * @desc Creates an action that updates the item's reference counter.
 * @param {Number} shift The value to add to the item's reference counter.
 *  Should be `1` or `-1` in majority of cases, although any other values
 *  are permitted.
 * @return {Action}
 */
function updateReferenceCounter(shift) {
  return shift;
}

export default redux.createActions({
  ITEM: {
    DROP_DATA: dropData,
    LOAD_DATA_INIT: loadDataInit,
    LOAD_DATA_DONE: loadDataDone,
    SET_DATA: setData,
    UPDATE_REFERENCE_COUNTER: updateReferenceCounter,
  },
});
