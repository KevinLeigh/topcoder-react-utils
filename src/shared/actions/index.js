import collection from './collection';
import item from './item';

export default {
  ...collection,
  ...item,
};
