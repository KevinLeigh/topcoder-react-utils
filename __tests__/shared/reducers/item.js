import actions from 'actions/item';
import reducer from 'reducers/item';
import mockDate from 'mockdate';

beforeAll(() => mockDate.set('2018-06-02T18:31:23+05:10'));
afterAll(() => mockDate.reset());

describe('Snapshot testing', () => {
  const a = actions.item;
  let state;

  function check(action) {
    state = reducer(state, action);
    expect(state).toMatchSnapshot();
    mockDate.set(Date.now() + 1234);
  }

  beforeEach(() => {
    state = undefined;
  });

  test('case #1', () => {
    check('@@INIT');
    check(a.updateReferenceCounter(1));
    check(a.setData('Test data'));
    check(a.dropData());
    check(a.updateReferenceCounter(-1));
    check(a.dropData());
  });

  test('case #2', () => {
    check('@@INIT');
    check(a.loadDataInit(12345));
    check(a.dropData());
    check(a.setData('Temporary Data'));
    check(a.loadDataDone(0, 'Wrong data'));
    check(a.loadDataDone(12345, 'Correct data'));
    check(a.dropData(Date.now() - 10000));
    check(a.updateReferenceCounter(-3));
    check(a.dropData());
  });
});
