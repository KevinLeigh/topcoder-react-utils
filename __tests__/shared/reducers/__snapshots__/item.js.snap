// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot testing case #1 1`] = `
Object {
  "data": null,
  "loadingOperationId": null,
  "numRefs": 0,
  "timestamp": 0,
}
`;

exports[`Snapshot testing case #1 2`] = `
Object {
  "data": null,
  "loadingOperationId": null,
  "numRefs": 1,
  "timestamp": 0,
}
`;

exports[`Snapshot testing case #1 3`] = `
Object {
  "data": "Test data",
  "loadingOperationId": null,
  "numRefs": 1,
  "timestamp": 1527945685468,
}
`;

exports[`Snapshot testing case #1 4`] = `
Object {
  "data": "Test data",
  "loadingOperationId": null,
  "numRefs": 1,
  "timestamp": 1527945685468,
}
`;

exports[`Snapshot testing case #1 5`] = `
Object {
  "data": "Test data",
  "loadingOperationId": null,
  "numRefs": 0,
  "timestamp": 1527945685468,
}
`;

exports[`Snapshot testing case #1 6`] = `
Object {
  "data": null,
  "loadingOperationId": null,
  "numRefs": 0,
  "timestamp": 0,
}
`;

exports[`Snapshot testing case #2 1`] = `
Object {
  "data": null,
  "loadingOperationId": null,
  "numRefs": 0,
  "timestamp": 0,
}
`;

exports[`Snapshot testing case #2 2`] = `
Object {
  "data": null,
  "loadingOperationId": 12345,
  "numRefs": 0,
  "timestamp": 0,
}
`;

exports[`Snapshot testing case #2 3`] = `
Object {
  "data": null,
  "loadingOperationId": 12345,
  "numRefs": 0,
  "timestamp": 0,
}
`;

exports[`Snapshot testing case #2 4`] = `
Object {
  "data": "Temporary Data",
  "loadingOperationId": 12345,
  "numRefs": 0,
  "timestamp": 1527945694106,
}
`;

exports[`Snapshot testing case #2 5`] = `
Object {
  "data": "Temporary Data",
  "loadingOperationId": 12345,
  "numRefs": 0,
  "timestamp": 1527945694106,
}
`;

exports[`Snapshot testing case #2 6`] = `
Object {
  "data": "Correct data",
  "loadingOperationId": null,
  "numRefs": 0,
  "timestamp": 1527945696574,
}
`;

exports[`Snapshot testing case #2 7`] = `
Object {
  "data": "Correct data",
  "loadingOperationId": null,
  "numRefs": 0,
  "timestamp": 1527945696574,
}
`;

exports[`Snapshot testing case #2 8`] = `
Object {
  "data": "Correct data",
  "loadingOperationId": null,
  "numRefs": -3,
  "timestamp": 1527945696574,
}
`;

exports[`Snapshot testing case #2 9`] = `
Object {
  "data": null,
  "loadingOperationId": null,
  "numRefs": -3,
  "timestamp": 0,
}
`;
